File: 面经150/3.滑动窗口/76.最小覆盖子串.py
#
# @lc app=leetcode.cn id=76 lang=python3
#
# [76] 最小覆盖子串
#
from mytools import *
# @lc code=start
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        hash = Counter(t)
        n = len(s); j = 0
        valid = 0
        sta = -1; l = inf
        for i in range(n):
            if s[i] in hash:
                hash[s[i]] -= 1
                if hash[s[i]] == 0: valid += 1
            while valid == len(hash):
                if i - j + 1 < l:
                    l = i - j + 1
                    sta = j
                if s[j] in hash:
                    hash[s[j]] += 1
                    if hash[s[j]] > 0: valid -= 1
                j += 1
        return s[sta: sta + l] if l != inf else ""
# @lc code=end

File: HOT100/HOT100p1/4.子串/76.最小覆盖子串.cpp
/*
 * @lc app=leetcode.cn id=76 lang=cpp
 *
 * [76] 最小覆盖子串
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    string minWindow(string s, string t) {
        int i = 0;
        unordered_map<char, int> hash;
        for (auto c : t)
        {
            ++ hash[c];
        }
        int valid = 0, minLen = INT32_MAX, start = -1;
        for (int j = 0; j < s.size(); j ++ )
        {
            if (hash.find(s[j]) != hash.end())
            {
                if(-- hash[s[j]] == 0) valid ++;
            }
            while (valid == hash.size())
            {
                if (j - i + 1 < minLen)
                {
                    start = i;
                    minLen = j - i + 1;
                }
                if (hash.find(s[i]) != hash.end())
                {
                    if (++ hash[s[i]] > 0) valid --;
                }
                i ++;
            }
        }
        return start == -1 ? "" : s.substr(start, minLen);
    }
};
// @lc code=end

